### PR TITLE
Basic Protocol Buffers support

### PR DESCRIPTION
**Pull request by soren121**
A critical feature of the hangups library is its use of [Protocol Buffers](https://developers.google.com/protocol-buffers/), Google's open interchange format for defining protocol serialization. Hangups uses a reverse-engineered implementation of Google's own protocol definition for Hangouts, and they make use of it pretty much *everywhere* in their library. Therefore, adding Protocol Buffers support is a crucial first step to porting the library over. Using PB will also enable us to merge changes from upstream easily.

A few problems exist with this in C# however:

 1. [Google's current Protocol Buffers library for C#](https://github.com/google/protobuf/) only supports the newer Proto3 language. However, since Hangouts uses Proto2 server-side, Hangups must also use Proto2 to comply. (see tdryer/hangups#170)
 2. There are two other options for C# ProtocolBuffers libraries: [mgravell/protobuf-net]( mgravell/protobuf-net) and [jskeet/protobuf-csharp-port](https://www.github.com/jskeet/protobuf-csharp-port).
    * protobuf-net does not support the PCL subset, so it won't work in a UWP project.
    * protobuf-csharp-port is a now-unsupported legacy library from Google that supports Proto2. Luckily for us, [phofman](https://www.github.com/phofman) ported it to WinRT/PCL, so we can use [his fork](https://github.com/phofman/protobuf-csharp-port).
 3. Unfortunately, neither of the above projects support the `oneof` field, which Hangups uses for notifications. If we want to support notifications, we'll have to patch in support for it in protobuf-csharp-port. Judging from how it's implemented in Google's Proto3 library, I don't think it would be terribly difficult.

Anyway, TL;DR: this PR adds the Proto2 definition file from Hangups and the C# code that I generated from it with ProtoGen, which is a tool included with protobuf-csharp-port that generates native code from .proto definitions.

I also included pre-built binaries of phofman's protobuf-csharp-port fork. I don't know of a better way to include it in this project; if anyone knows, please let me know!

### CODE REVIEW

**Comment by kfechter** [comment]
sorry I haven't been working on it as much. Just finished school and am spending a bit of time with families. now that we have protobuf, I'll spend some time looking at it.

**Comment by soren121** [comment]
It's totally fine! Don't feel like you're obligated to work on your hobby projects just because you think people expect it of you-- that's when it becomes less fun to work on them.

**Comment by kfechter** [comment]
Alright! I'll keep that in mind.


Also, I noticed that adding protobuf requires a reference to .NET 4.0, but not sure that can be added with a universal app.

**Comment by soren121** [comment]
That's strange, it shouldn't. I built it to target Win8 and WP8.1, but not .NET 4.0. I'll clone it on my other PC and check it out.

**Comment by kfechter** [comment]
ahh. I'm targeting windows 10 UWP with mine.

**Comment by soren121** [comment]
I know. When I tried to set it to target UWP, Visual Studio told me that I should build it for Win8.1 because the Win8.1 API is a subset of UWP and the protobuf project doesn't use any UWP-specific API's,

**Comment by soren121** [comment]
Also, this is embarrassing but it looks like I've never tried to build Hangups yet.

I don't want to put in a PR just for this, but you need to find-and-replace all instances of `Hangups.Core` with `Hangups.Proto` in Proto/Proto.cs to get it to build. I forgot to do that beforehand.

**Comment by kfechter** [comment]
interesting. also, yeah I changed that. Maybe I missed them somewhere else.

**Comment by soren121** [comment]
Here's a screenshot of my references in VS. Does it look different on your PC?

capture

**Comment by kfechter** [comment]
Nope, exact same. 

works now though.
